File: packages/apple/Example/LocanaraExample/components/pages/FrameworkShowcase/ModelDemo.swift
import SwiftUI
import Locanara

/// Demonstrates LocanaraModel — direct model usage with GenerationConfig presets and streaming
struct ModelDemo: View {
    @EnvironmentObject var appState: AppState
    @State private var inputText = "Explain what on-device AI means in one sentence."
    @State private var selectedPreset = ConfigPreset.conversational
    @State private var resultText: String?
    @State private var processingTime: Int?
    @State private var isLoading = false
    @State private var errorMessage: String?
    @State private var useStreaming = false
    @State private var streamedText = ""

    private enum ConfigPreset: String, CaseIterable, Identifiable {
        case structured
        case creative
        case conversational

        var id: String { rawValue }

        var title: String {
            switch self {
            case .structured: return "Structured"
            case .creative: return "Creative"
            case .conversational: return "Conversational"
            }
        }

        var description: String {
            switch self {
            case .structured: return "temp 0.2, topK 16 — precise, factual"
            case .creative: return "temp 0.8, topK 40 — varied, expressive"
            case .conversational: return "temp 0.7, topK 40 — natural, balanced"
            }
        }

        var config: GenerationConfig {
            switch self {
            case .structured: return .structured
            case .creative: return .creative
            case .conversational: return .conversational
            }
        }
    }

    private var isAIAvailable: Bool {
        appState.currentEngine != .none && appState.isModelReady
    }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 20) {
                if !isAIAvailable {
                    AIModelRequiredBanner()
                }

                CodePatternView(code: """
                let model = FoundationLanguageModel()

                // Direct generation with config presets
                let response = try await model.generate(
                    prompt: "Your prompt here",
                    config: .\(selectedPreset.rawValue)
                )
                print(response.text)
                print(response.processingTimeMs)

                // Streaming
                for try await chunk in model.stream(prompt: "...") {
                    print(chunk)
                }
                """)

                VStack(alignment: .leading, spacing: 8) {
                    Text("Prompt")
                        .font(.headline)
                    TextEditor(text: $inputText)
                        .frame(minHeight: 100)
                        .padding(8)
                        .background(Color.gray.opacity(0.15))
                        .clipShape(RoundedRectangle(cornerRadius: 8))
                        .disabled(isLoading)
                }

                VStack(alignment: .leading, spacing: 8) {
                    Text("GenerationConfig Preset")
                        .font(.headline)
                    Picker("Preset", selection: $selectedPreset) {
                        ForEach(ConfigPreset.allCases) { preset in
                            Text(preset.title).tag(preset)
                        }
                    }
                    .pickerStyle(.segmented)

                    Text(selectedPreset.description)
                        .font(.caption)
                        .foregroundStyle(.secondary)
                }

                Toggle("Use Streaming", isOn: $useStreaming)
                    .font(.headline)

                Button(action: executeModel) {
                    HStack(spacing: 8) {
                        if isLoading {
                            ProgressView()
                                .scaleEffect(0.8)
                                .tint(.white)
                        }
                        Text(isLoading ? "Generating..." : "Generate")
                    }
                    .frame(maxWidth: .infinity)
                    .frame(height: 20)
                }
                .buttonStyle(.borderedProminent)
                .disabled(isLoading || inputText.isEmpty || !isAIAvailable)

                if let errorMessage {
                    Text(errorMessage)
                        .foregroundStyle(.red)
                        .font(.caption)
                }

                if useStreaming && !streamedText.isEmpty {
                    VStack(alignment: .leading, spacing: 8) {
                        Label("Streamed Output", systemImage: "text.line.first.and.arrowtriangle.forward")
                            .font(.headline)
                            .foregroundStyle(.blue)
                        Text(streamedText)
                            .padding(12)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .background(Color.blue.opacity(0.1))
                            .clipShape(RoundedRectangle(cornerRadius: 8))
                            .animation(.easeInOut(duration: 0.1), value: streamedText)
                    }
                }

                if let resultText {
                    VStack(alignment: .leading, spacing: 8) {
                        Label("Model Response", systemImage: "cpu")
                            .font(.headline)
                            .foregroundStyle(.green)
                        Text(resultText)
                            .padding(12)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .background(Color.green.opacity(0.1))
                            .clipShape(RoundedRectangle(cornerRadius: 8))

                        if let processingTime {
                            HStack(spacing: 12) {
                                StatBadge(label: "Time", value: "\(processingTime)ms")
                                StatBadge(label: "Config", value: selectedPreset.title)
                            }
                        }
                    }
                }
            }
            .padding()
        }
    }

    private func executeModel() {
        isLoading = true
        errorMessage = nil
        resultText = nil
        streamedText = ""
        processingTime = nil

        Task {
            do {
                let model = FoundationLanguageModel()
                let config = selectedPreset.config

                if useStreaming {
                    let startTime = Date()
                    var accumulated = ""
                    for try await chunk in model.stream(prompt: inputText, config: config) {
                        accumulated += chunk
                        await MainActor.run { self.streamedText = accumulated }
                    }
                    let elapsed = Int(Date().timeIntervalSince(startTime) * 1000)
                    await MainActor.run {
                        self.resultText = accumulated
                        self.processingTime = elapsed
                    }
                } else {
                    let response = try await model.generate(prompt: inputText, config: config)
                    await MainActor.run {
                        self.resultText = response.text
                        self.processingTime = response.processingTimeMs
                    }
                }

            } catch {
                await MainActor.run { errorMessage = error.localizedDescription }
            }
            await MainActor.run { isLoading = false }
        }
    }
}

File: packages/android/example/src/main/kotlin/com/locanara/example/components/pages/framework/ModelDemo.kt
package com.locanara.example.components.pages.framework

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Memory
import androidx.compose.material3.Button
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.FilterChip
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import com.locanara.core.GenerationConfig
import com.locanara.core.ModelResponse
import com.locanara.platform.PromptApiModel
import kotlinx.coroutines.launch

private enum class ConfigPreset(val label: String, val description: String) {
    STRUCTURED("Structured", "temp 0.2, topK 16 — precise"),
    CREATIVE("Creative", "temp 0.8, topK 40 — varied"),
    CONVERSATIONAL("Conversational", "temp 0.7, topK 40 — natural")
}

/**
 * Demonstrates LocanaraModel — direct model usage with GenerationConfig presets and streaming.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ModelDemo(onNavigateBack: () -> Unit) {
    var inputText by remember { mutableStateOf("Explain what on-device AI means in one sentence.") }
    var selectedPreset by remember { mutableStateOf(ConfigPreset.CONVERSATIONAL) }
    var useStreaming by remember { mutableStateOf(false) }
    var resultText by remember { mutableStateOf<String?>(null) }
    var streamedText by remember { mutableStateOf("") }
    var processingTime by remember { mutableStateOf<Int?>(null) }
    var isLoading by remember { mutableStateOf(false) }
    var errorMessage by remember { mutableStateOf<String?>(null) }
    val scope = rememberCoroutineScope()
    val context = LocalContext.current

    val codePattern = """
val model = PromptApiModel(context)

// Direct generation with config presets
val response = model.generate(
    prompt = "Your prompt here",
    config = GenerationConfig.${selectedPreset.name}
)
println(response.text)
println(response.processingTimeMs)

// Streaming
model.stream(prompt = "...").collect { chunk ->
    print(chunk)
}
    """.trimIndent()

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Model") },
                navigationIcon = {
                    IconButton(onClick = onNavigateBack) {
                        Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = "Back")
                    }
                }
            )
        }
    ) { padding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
                .padding(horizontal = 16.dp)
                .verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(16.dp)
        ) {
            CodePatternCard(code = codePattern)

            OutlinedTextField(
                value = inputText,
                onValueChange = { inputText = it },
                label = { Text("Prompt") },
                modifier = Modifier.fillMaxWidth(),
                minLines = 3,
                enabled = !isLoading
            )

            Column {
                Text("GenerationConfig Preset", style = MaterialTheme.typography.titleSmall, fontWeight = FontWeight.Medium)
                Spacer(modifier = Modifier.height(4.dp))
                Text(selectedPreset.description, style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
                Spacer(modifier = Modifier.height(8.dp))
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    ConfigPreset.entries.forEach { preset ->
                        FilterChip(
                            selected = selectedPreset == preset,
                            onClick = { selectedPreset = preset },
                            label = { Text(preset.label) }
                        )
                    }
                }
            }

            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                Text("Use Streaming", style = MaterialTheme.typography.titleSmall)
                Switch(checked = useStreaming, onCheckedChange = { useStreaming = it })
            }

            Button(
                onClick = {
                    isLoading = true
                    errorMessage = null
                    resultText = null
                    streamedText = ""
                    processingTime = null
                    scope.launch {
                        try {
                            println("[ModelDemo] input: ${inputText.take(200)}, config: ${selectedPreset.label}, streaming: $useStreaming")
                            val model = PromptApiModel(context)
                            val config = when (selectedPreset) {
                                ConfigPreset.STRUCTURED -> GenerationConfig.STRUCTURED
                                ConfigPreset.CREATIVE -> GenerationConfig.CREATIVE
                                ConfigPreset.CONVERSATIONAL -> GenerationConfig.CONVERSATIONAL
                            }

                            if (useStreaming) {
                                val startTime = System.currentTimeMillis()
                                var accumulated = ""
                                model.stream(prompt = inputText, config = config).collect { chunk ->
                                    accumulated += chunk
                                    streamedText = accumulated
                                }
                                processingTime = (System.currentTimeMillis() - startTime).toInt()
                                resultText = accumulated
                            } else {
                                val response = model.generate(prompt = inputText, config = config)
                                resultText = response.text
                                processingTime = response.processingTimeMs
                            }
                            println("[ModelDemo] result: ${resultText?.take(200)}, time: ${processingTime}ms")
                        } catch (e: Exception) {
                            println("[ModelDemo] error: ${e.message}")
                            errorMessage = e.message ?: "Unknown error"
                        } finally {
                            isLoading = false
                        }
                    }
                },
                modifier = Modifier.fillMaxWidth(),
                enabled = !isLoading && inputText.isNotBlank()
            ) {
                if (isLoading) {
                    CircularProgressIndicator(modifier = Modifier.padding(end = 8.dp), strokeWidth = 2.dp)
                }
                Text(if (isLoading) "Generating..." else "Generate")
            }

            errorMessage?.let { error ->
                Text(error, color = MaterialTheme.colorScheme.error, style = MaterialTheme.typography.bodySmall)
            }

            if (useStreaming && streamedText.isNotEmpty()) {
                Card(
                    colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.primaryContainer),
                    shape = RoundedCornerShape(12.dp)
                ) {
                    Column(modifier = Modifier.padding(16.dp)) {
                        Text("Streamed Output", style = MaterialTheme.typography.titleSmall, fontWeight = FontWeight.Medium)
                        Spacer(modifier = Modifier.height(8.dp))
                        Text(streamedText, style = MaterialTheme.typography.bodyMedium)
                    }
                }
            }

            resultText?.let { result ->
                Card(
                    colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.secondaryContainer),
                    shape = RoundedCornerShape(12.dp)
                ) {
                    Column(modifier = Modifier.padding(16.dp)) {
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Icon(Icons.Default.Memory, contentDescription = null, tint = MaterialTheme.colorScheme.primary)
                            Text("  Model Response", style = MaterialTheme.typography.titleSmall, fontWeight = FontWeight.Medium)
                        }
                        Spacer(modifier = Modifier.height(8.dp))
                        Text(result, style = MaterialTheme.typography.bodyMedium)
                    }
                }

                processingTime?.let { time ->
                    Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
                        StatBadge(label = "Time", value = "${time}ms")
                        StatBadge(label = "Config", value = selectedPreset.label)
                    }
                }
            }

            Spacer(modifier = Modifier.height(16.dp))
        }
    }
}
